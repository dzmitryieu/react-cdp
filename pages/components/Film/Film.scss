@import "../../../src/utils/constants.scss";

.container {
  justify-content: center;
  display: flex;
  box-sizing: border-box;
  font-family: $helvetica;
  margin-top: 20px;

  .wrapper {
    height: 100%;
    width: 300px;
    position: relative;
    cursor: pointer;
    transition: all ease 0.3s;

    &:hover {
      transform: translate(3px 3px);
      box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
    }

    .poster {
      background-size: cover;
      background-position: center center;
      background-repeat: no-repeat;
      height: 370px;
      text-align: center;
    }
  }
}

.link {
  text-decoration: none;
}
