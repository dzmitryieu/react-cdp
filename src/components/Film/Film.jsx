import React from 'react';
import PropTypes from 'prop-types';
import styles from './Film.scss';
import FilmInfo from '../FilmInfo';
import defaultPoster from '../../assets/images/default-poster.jpg';

class Film extends React.Component {
  constructor(props) {
    super(props);
    this.myRef = React.createRef();
  }

  componentDidMount() {
    const { onClick } = this.props;
    const container = this.myRef.current;
    container.addEventListener('click', onClick);
  }

  componentWillUnmount() {
    const { onClick } = this.props;
    const container = this.myRef.current;
    container.removeEventListener('click', onClick);
  }

  render() {
    const {
      name,
      posterPath,
      genres,
      year,
    } = this.props;
    const posterURL = posterPath || defaultPoster;
    return (
      <div id="film" className={styles.container} ref={this.myRef}>
        <div className={styles.wrapper}>
          <div
            className={styles.poster}
            style={{
              backgroundImage: `url(${posterURL})`,
            }}
          />
          <FilmInfo
            name={name}
            genres={genres}
            year={year}
          />
        </div>
      </div>
    );
  }
}

Film.propTypes = {
  name: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  posterPath: PropTypes.string,
  onClick: PropTypes.func.isRequired,
};
Film.defaultProps = {
  posterPath: null,
};

export default Film;
