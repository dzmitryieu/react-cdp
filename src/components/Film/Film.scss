@import "../../utils/constants.scss";

.container {
  justify-content: center;
  display: flex;
  box-sizing: border-box;
  font-family: $helvetica;
  margin-top: 20px;

  .wrapper {
    height: 450px;
    width: 300px;
    position: relative;
    cursor: pointer;
    transition: all ease 0.3s;

    &:hover {
      transform: translate(3px 3px);
      box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
    }

    .poster {
      background-size: cover;
      background-position: center center;
      background-repeat: no-repeat;
      height: 370px;
      text-align: center;

      .hover-screen {
        display: none;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        background-color: rgba(0, 0, 0, 0.4);
        height: 100%;
        transition: all 1s ease;
      }

      .play {
        width: 60px;
        height: 60px;
        background-color: $light-red;
        border-radius: 50%;
        position: relative;
        cursor: pointer;
        transition: all 0.3s linear;

        &::after {
          content: '';
          display: block;
          top: 16px;
          left: 23px;
          width: 25px;
          height: 25px;
          border: 14px solid transparent;
          border-left: 20px solid white;
          box-sizing: border-box;
          position: absolute;
          z-index: 0;
        }

        &:hover {
          transform: scale(1.1);
        }
      }

      .signature {
        font-family: $helvetica;
        font-size: 14px;
        text-align: center;
        color: white;
      }
    }
  }
}

.info-wrapper {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  z-index: 20;

  .shadow {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    position: absolute;
    top: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.7);
    z-index: 30;
    height: 100%;
    width: 100%;
    padding: 10px 20px;
    box-sizing: border-box;

    .cross-wrapper {
      width: 100%;
      height: 50px;
      display: flex;
      justify-content: flex-start;
      align-items: center;
      box-sizing: border-box;

      .cross {
        color: white;
        background-color: transparent;
        border: none;
        outline: none;
        font-size: 25px;
        cursor: pointer;
        margin: 0;
        transition: all 0.3s ease;
        z-index: 120;
        position: relative;

        &:hover {
          transform: scale(1.1);
        }
      }
    }

    .film-description {
      font-size: 15px;
      line-height: 24px;
      color: white;
      font-family: $helvetica;
      text-align: left;
      font-weight: 100;
      max-height: 295px;
      overflow: hidden;
    }

    button {
      margin-left: 0;
    }
  }
}
