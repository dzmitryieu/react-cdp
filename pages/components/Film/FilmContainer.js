import { withRouter } from 'react-router-dom';

import Film from './Film';

export default withRouter(Film);
