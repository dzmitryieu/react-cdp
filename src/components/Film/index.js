import Film from './FilmContainer';

export default Film;
