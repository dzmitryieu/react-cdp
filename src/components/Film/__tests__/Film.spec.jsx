import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import Film from '../index';

const renderer = new ShallowRenderer();

const onClick = jest.fn();

const film = {
  id: '13456',
  name: 'Captain Marvel',
  poster_path: '../../assets/images/poster.jpg',
  genres: ['Drama', 'Comedy'],
  year: '1984',
};

const film2 = {
  id: '13456',
  name: 'Captain Marvel',
  genres: ['Drama', 'Comedy'],
  year: '1984',
};

describe('Film Snapshot', () => {
  test('renders', () => {
    const component = renderer.render(
      <Film {...film} onClick={onClick} />,
    );
    expect(component).toMatchSnapshot();
  });
  test('renders without PosterPath', () => {
    const component = renderer.render(
      <Film {...film2} onClick={onClick} />,
    );
    expect(component).toMatchSnapshot();
  });
});

describe('Film lifecycle method componentWillUnmount should have been called', () => {
  test('lifecycle method should have been called', () => {
    const component = TestRenderer.create(
      <Film {...film} onClick={onClick} />,
      {
        createNodeMock: (element) => {
          if (element.props.className === 'container') {
            return {
              addEventListener: () => {},
              removeEventListener: () => {},
            };
          }
          return null;
        },
      },
    );
    component.unmount();
  });
});
