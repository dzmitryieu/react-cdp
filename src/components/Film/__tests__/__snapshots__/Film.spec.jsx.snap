// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film Snapshot renders 1`] = `
<div
  className="container"
  id="film"
>
  <div
    className="wrapper"
  >
    <div
      className="poster"
      style={
        Object {
          "backgroundImage": "url(test-file-stub)",
        }
      }
    />
    <FilmInfo
      genres={
        Array [
          "Drama",
          "Comedy",
        ]
      }
      name="Captain Marvel"
      year="1984"
    />
  </div>
</div>
`;

exports[`Film Snapshot renders without PosterPath 1`] = `
<div
  className="container"
  id="film"
>
  <div
    className="wrapper"
  >
    <div
      className="poster"
      style={
        Object {
          "backgroundImage": "url(test-file-stub)",
        }
      }
    />
    <FilmInfo
      genres={
        Array [
          "Drama",
          "Comedy",
        ]
      }
      name="Captain Marvel"
      year="1984"
    />
  </div>
</div>
`;
